### PR TITLE
[Best Practices] Database Sessions

### PR DESCRIPTION
**Pull request by irmiller22**
## Background

Add document detailing specifics around database sessions, transactions, and guidance around best practices.

### CODE REVIEW

**Comment by irmiller22** [comment]
@nandadong bump

File: best-practices/databases/session-and-transactions.md [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+- [What is a database transaction?](#what-is-a-database-transaction)
+  - [Explain it to me like I'm 5](#explain-it-to-me-like-im-5)
+- [SQLAlchemy](#sqlalchemy)
+  - [How does SQLAlchemy leverage sessions and transactions?](#how-does-sqlalchemy-leverage-sessions-and-transactions)
+- [Best Practices](#best-practices)
+- [Resources](#resources)
+
+## Introduction
+
+The purpose of this document is to introduce three tpics:
```
Comment (nandadong) [inline best-practices/databases/session-and-transactions.md]:
```suggestion
The purpose of this document is to introduce three topics:
```

File: best-practices/databases/session-and-transactions.md [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+
+- overview of database sessions and database transactions
+- how sessions and transactions are managed via SQLAlchemy
+- best practices for managing database sessions and transactions
+
+The itention is to come away from this document with knowledge on how to manage database sessions and transactions in the backend codebase.
```
Comment (nandadong) [inline best-practices/databases/session-and-transactions.md]:
```suggestion
The intention is to come away from this document with knowledge on how to manage database sessions and transactions in the backend codebase.
```

File: best-practices/databases/session-and-transactions.md [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+
+A database session is the code construct that establishes and maintains all communication between your application and the database(s). It is intermediary in which all application logic and database abstractions (SQLAlchemy model, for example) are loaded. Most importantly, it manages database transactions initiated in the form of queries.
+
+A database session lifecycle looks like the following:
+
+1. Database session is created.
+1. Database session retrieves query requests, and the query results are associated with the database session.
+1. Database model objects are added to the database session.
+1. Database session starts to manage the changes requested against the database model objects.
+1. Database session commits the changes (if commit fails, then a rollback occurs).
```
Comment (nandadong) [inline best-practices/databases/session-and-transactions.md:34]: Should we also cover `session.flush` somewhere?

File: best-practices/databases/session-and-transactions.md [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+
+There are 3 states associated with transactions: `active`, `idle in transaction`, and `idle`. Most are familiar with the `active` state - this is simply a database connection that is doing work.
+
+The other two states, `idle` and `idle in transaction` are important to keep in mind for debugging purposes. The `idle` state means that the database connection is not doing anything, and is waiting. This is a good point to disconnect from the database since this won't result in data loss. The `idle in transaction` state means that the connection started a transaction at some point, and is waiting for something. This can be anything like data being processed on the application side, the database loading some data before returning it to the application, or something along those lines. Whenever you have connections that are in the `idle in transaction` state, this usually means that your database connection setup warrants more investigation, because the connections are not being used efficiently.
+
+### Explain it to me like I'm 5.
```
Comment (nandadong) [inline best-practices/databases/session-and-transactions.md:49]: 👶

File: best-practices/databases/session-and-transactions.md [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+
+Imagine that you're going to a bank. You're planning to deposit 3 checks, and also withdraw some cash.
+
+You walk into the bank, and you wait in line until a teller window opens up. You fill out the deposit slips, and then deposit the checks one by one. After each deposit, the teller gives you confirmation that the check has been deposited, as well as the amount that has changed in your bank account. Finally, you request the cash withdrawal. The teller hands you the cash.
+
+Once you're finished, you step out of the line, and walk out of the bank.
+
+Imagine that walking into the bank is like creating a session, and walking out of the bank is like closing a session. Waiting in the teller line for an available teller is analogous to waiting for an available connection via the connection pool. Handing over each check to be deposited and retrieving confirmations are database transactions. Finally, asking for and retrieving the cash is a separate database transactions.
+
```
Comment (nandadong) [inline best-practices/databases/session-and-transactions.md:58]: Great explanation!

**Review by nandadong — APPROVED** [review]
Awesome! 👍

File: best-practices/databases/session-and-transactions.md (hunk 3 above)
Comment (irmiller22) [reply best-practices/databases/session-and-transactions.md:34]: @nandadong by default, `commit` calls `flush`. Might be worthwhile to cover this separately. For example, discuss how flush can be used for non-default behavior (like in maintaining order of operations for queries - `commit` doesn't guarantee this).

Comment (irmiller22) [reply best-practices/databases/session-and-transactions.md:34]: Added here: https://github.com/irmiller22/tamagotchi/pull/5/commits/e05b67c28156413b1e81f438d08cff50b0d88efb

Comment (nandadong) [reply best-practices/databases/session-and-transactions.md:34]: 🔥

**Pull request merged**